### ISSUE TITLE
Some troubles with a small example using the Boost C++ library

### ISSUE THREAD

**Issue by vincent-picaud** [verificarlo/verificarlo#44]
Hi,

First of all thank you for sharing Verificarlo, that is really a promising project!

I just successfully installed it using your debian package

I also run with success the tests (test_tchebychev, test_...) that come with the source. 

I just tried something more complicated, a special function of the Boost C++ library, but I got some troubles...

I wanted to test the  [Boost digamma function](http://www.boost.org/doc/libs/1_63_0/libs/math/doc/html/math_toolkit/sf_gamma/digamma.html)

My small C++ code is

```
#include <boost/math/special_functions/digamma.hpp>
#include <iostream>

int
main()
{
  std::cout << "\n result " << boost::math::digamma(0.1);
}
```
I try to compile it with:

`verificarlo oneExample.cpp -o oneExample`

But I get the following error messages:

```
Unsupported operand type: x86_fp80
Unsupported operand type: x86_fp80
Unsupported operand type: x86_fp80
Unsupported operand type: x86_fp80
Unsupported operand type: x86_fp80
Unsupported operand type: x86_fp80
Unsupported operand type: x86_fp80
Unsupported operand type: x86_fp80
Call parameter type does not match function signature!
x86_fp80 0xK3FFF8000000000000000
 float  %11 = call float %10(x86_fp80 0xK3FFF8000000000000000, x86_fp80 %9)
Stored value type does not match pointer operand type!
  store float %11, x86_fp80* %2, align 16
 x86_fp80Call parameter type does not match function signature!
  %12 = load x86_fp80, x86_fp80* %2, align 16
 float  %16 = call float %15(x86_fp80 %12, x86_fp80 %14)
Stored value type does not match pointer operand type!
  store float %16, x86_fp80* %remainder, align 16
...
 x86_fp80LLVM ERROR: Broken function found, compilation aborted!
/usr/bin/verificarlo: command failed:
/usr/lib/llvm-3.7/bin/opt -S  -load /usr/lib/x86_64-linux-gnu/libvfcinstrument.so -vfclibinst   oneExample.1.ll -o oneExample.2.ll
```

My question, as a new  Verificarlo user, is:
1/ Do I missed something (in my compile command etc...)?
or
2/ Verificarlo is not yet fully fonctional and I have to wait until it supports libs like Boost?
or
3/ Is it a Verificarlo bug?

- Vincent

ps: I also posted this question in the [google group](https://groups.google.com/forum/#!topic/verificarlo/axfHbmUsrMo) but maybe here is a better place.

**Comment by pablooliveira** [comment]
Dear Vincent,

On 03/15/2017 12:15 PM, Vincent Picaud wrote:
> Hi,
>
> First of all thank you for sharing Verificarlo, that is really a
> promising project!

Thanks for your interest in Verificarlo. I'm glad you could install and
try it.

> [...]
> However I have troubles with this example that uses the digamma function
> from the boost library:
> http://www.boost.org/doc/libs/1_63_0/libs/math/doc/html/math_toolkit/sf_gamma/digamma.html
> Unsupported operand type: x86_fp80
> [...]
> My question is:
> 1/ Do I missed something (in my compile command etc...)?
> or
> 2/ Verificarlo is not yet fully fonctional and I have to wait until it
> supports lib like Boost?
> or
> 3/ Is it a Verificarlo bug?

Your command is correct. The problem here is with the boost library
generating code that relies on x87 80 bits extended precision.

Unfortunately, we have not yet added x87 extended fp80 support in
verificarlo. Thanks for providing a small failing example.

In the meantime, a work around for this is to pass the -mno-x87 option
to verificarlo. This option forbids usage of x87 instructions. The
problem is that the option -mno-x87 has been added recently to LLVM 3.9.
Right now verificarlo only support LLVM 3.8.1, so the option does nor
work...

I'm a bit busy right now, but I will try soon to upgrade verificarlo to
work with LLVM 3.9. I'll keep you posted.

Thanks for this bug report !

Pablo

**Comment by vincent-picaud** [comment]
Bonjour Pablo,

Merci de votre réponse. J'ai eu aussi une réponse de Christophe. Afin de me
situer un peu je me permets un copié/collé:

""
Je suis à l'étage au dessus de la maison de la simulation (CEA/Saclay).
Je n'ai malheureusement pas pu assister à la présentation de Christophe,
mais j'en ai eu connaissance et c'est ainsi que j'ai appris l'existence de
Verificarlo.

Ici au CEA je fais plutot du traitement de données, mais j'ai aussi un
passé plus orienté calcul par éléments finis.

Je trouve l'approche de Verificarlo très intéressante et ma démarche est
une démarche de "curiosité personnelle" (je veux dire par là que je ne suis
pas un riche industriel :-) ).

Ceci étant, je contribue au projet Stan où il semblerait qu'il y ai des
difficultés lié aux problèmes de précision des calculs:
http://discourse.mc-stan.org/t/numerical-errors-do-we-have-some-intuition-on-its-severity/297/5
http://discourse.mc-stan.org/t/numerical-error-tanks-stepsize-other-possibilities/293

Mon idée était d'utiliser Verficarlo dans ce cadre. Cependant le code C++
de Stan utilise Boost, donc la première étape était de regarder à ce
niveau, d'où mon post
""

Your command is correct. The problem here is with the boost library
> generating code that relies on x87 80 bits extended precision.
>
 Merci pour cet éclaircissement.

> Unfortunately, we have not yet added x87 extended fp80 support in
> verificarlo. Thanks for providing a small failing example.
>
J'imagine que c'est énormément de travaille, mais à l'avenir qu'elles sont
vos intentions?
Vous pensez étendre Verificarlo pour supporter la représentation interne
Intel ou pas?

> In the meantime, a work around for this is to pass the -mno-x87 option
> to verificarlo. This option forbids usage of x87 instructions. The
> problem is that the option -mno-x87 has been added recently to LLVM 3.9.
> Right now verificarlo only support LLVM 3.8.1, so the option does nor
> work...
>
Je ne connaissais pas ce flag de compilation. Mais juste pour info il cause
un crash de clang++ quand j'essaie de compiler mon exemple.
 $ clang++ --version
clang version 3.9.1-4 (tags/RELEASE_391/rc2)
Target: x86_64-pc-linux-gnu


I'm a bit busy right now, but I will try soon to upgrade verificarlo to
> work with LLVM 3.9. I'll keep you posted.
>
> Thanks for this bug report !
>
Je comprends aucun problème, je suis moi même pris par des projets en
souffrances :-(.

Comme je disais à Christophe, je n'ai aucune connaissance en théorie de la
compilation/LLVM je ne peux donc pas vraiment aider.
Je peux ajouter peut être un ou deux exemples d'utilisation: dérivée par
différence finie ou extrapolation de richardson.

Merci pour votre réponse
Vincent

2017-03-16 12:09 GMT+01:00 Pablo de Oliveira Castro <
notifications@github.com>:

**Comment by vincent-picaud** [comment]
Here is a feedback concerning my problem.

Boost::math has a macro, BOOST_MATH_NO_LONG_DOUBLE_MATH_FUNCTIONS, that disables long double computations.

Now my code compiles with it:

> verificarlo -lm -lstdc++ -DBOOST_MATH_NO_LONG_DOUBLE_MATH_FUNCTIONS oneExample.cpp -o oneExample

Unfortunately I get a SIGSEGV at running time... I have not investigated this (there is no SIGSEGV when compiling with g++ or clang++).

However for another special functions, like the Gamma one

> #include <boost/math/special_functions/gamma.hpp>
> #include <iostream>
> 
> int
> main()
> {
>   std::cout << "\n result " << boost::math::tgamma(0.1);
> }

everything is running smoothly (-DBOOST_MATH_NO_LONG_DOUBLE_MATH_FUNCTIONS being still required in the compilation command line).

-Vincent

**Comment by pablooliveira** [comment]
Thanks a lot for the reporting your work-around for long double Vincent.

It would be interesting to see why verificarlo makes your program segfault. Could you provide us a stack-trace of the crash ?

One way to get it should be something like:
```
$ ulimit -c unlimited
$ ./application
$ gdb ./application core
>> bt
```

**Comment by vincent-picaud** [comment]
Sure

Program received signal SIGSEGV, Segmentation fault.
0x0000000000000000 in ?? ()
(gdb) bt
#0  0x0000000000000000 in ?? ()
#1  0x00000000004040ac in boost::math::detail::digamma_imp_1_2<double> (x=1.5) at /usr/include/boost/math/special_functions/digamma.hpp:337
#2  0x0000000000402498 in boost::math::detail::digamma_imp<double, mpl_::int_<53>, boost::math::policies::policy<boost::math::policies::promote_float<false>, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy> > (x=1.5, t=0x0, pol=...) at /usr/include/boost/math/special_functions/digamma.hpp:447
#3  0x0000000000405568 in boost::math::digamma<double, boost::math::policies::policy<boost::math::policies::promote_float<false>, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy> > (x=1.5) at /usr/include/boost/math/special_functions/digamma.hpp:615
#4  0x0000000000405515 in boost::math::detail::digamma_initializer<double, boost::math::policies::policy<boost::math::policies::promote_float<false>, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy> >::init::do_init (this=0x6093dd <boost::math::detail::digamma_initializer<double, boost::math::policies::policy<boost::math::policies::promote_float<false>, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy> >::initializer>) at /usr/include/boost/math/special_functions/digamma.hpp:559
#5  0x0000000000402089 in boost::math::detail::digamma_initializer<double, boost::math::policies::policy<boost::math::policies::promote_float<false>, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy> >::init::init (this=0x6093dd <boost::math::detail::digamma_initializer<double, boost::math::policies::policy<boost::math::policies::promote_float<false>, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy, boost::math::policies::default_policy> >::initializer>) at /usr/include/boost/math/special_functions/digamma.hpp:555
#6  0x0000000000401ee1 in __cxx_global_var_init.1(void) () at /usr/include/boost/math/special_functions/digamma.hpp:573
#7  0x000000000040615d in __libc_csu_init ()
#8  0x00007ffff6db3240 in __libc_start_main (main=0x401ff0 <main()>, argc=1, argv=0x7fffffffe268, init=0x406110 <__libc_csu_init>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffe258) at ../csu/libc-start.c:247
#9  0x0000000000401f1a in _start () at /usr/include/boost/math/special_functions/digamma.hpp:573

My config:
- verificarlo: debian package

> verificarlo --version
> verificarlo 0.2.0

- boost: BOOST_LIB_VERSION "1_62" (debian testing version)

- my code 

> #include <boost/math/special_functions/digamma.hpp>
> #include <iostream>
> 
> int
> main()
> {
>   std::cout << "\n result " << boost::math::digamma(0.1);
> }

- my compilation command line

> verificarlo -g -lm -lstdc++ -DBOOST_MATH_NO_LONG_DOUBLE_MATH_FUNCTIONS oneExample.cpp -o oneExample

Hope that you have everything
-Vincent

**Comment by vincent-picaud** [comment]
(the second #include is #include <iostream>, I do not know why this was rubbed out)

**Comment by pablooliveira** [comment]
Dear Vincent,

I found the root cause of this ! Thanks for your valuable info. The fix is at #50 and should be merged soon :tada:

**Comment by vincent-picaud** [comment]
Dear Pablo,
Wow, that is a great news!
Thank you for your work,
Vincent

2017-05-03 14:37 GMT+02:00 Pablo de Oliveira Castro <
notifications@github.com>:
